### DISCUSSION TITLE
Do tools (function calling) work across LLMs?

### DISCUSSION THREAD

**Discussion by rwankar** [langchain4j/langchain4j#1847] [Q&A]
Hello,

I built an example with two functions with gpt-4o-mini and LangChain4j. This works great. So I thought to experiment with other LLMs to see how they compare. I opened an account with Groq so I can play around with Llama and a few other LLMs. To connect to Groq I did the following.

```
assistant = AiServices.builder(Assistant.class)
                .chatLanguageModel(OpenAiChatModel.builder()
                    .baseUrl("https://api.groq.com/openai/v1")
                    .apiKey(ApiKeys.GROQ_API_KEY)
                    .modelName(model)
                    .logRequests(true)
                    .logResponses(true)
                    .build())
                .tools(new MyTools())
                .chatMemoryProvider(memoryId -> TokenWindowChatMemory.withMaxTokens(1024, new OpenAiTokenizer()))
                .build();

```

I have 2 functions FindAuthor and FindBook.

The FindAuthor tool takes 2 arguments
- bookTitle : a string to find author by the book title
- authorName : a string to find author by name

The FindBooks function takes 2 arguments
- bookTitle : a string to find a book by title
- authorId : an array of ids for the author

I put in a system prompt explaining the job.

My prompt was **Which other books were written by the author of Angels and Daemons.**

My expectation is that the LLM should first call the FindAuthor tool with the bookTitle argument which returns a JSON object with  Dan Brown and the id. Then the LLM should call FindBooks with the authorId as the argument.

gpt-4o-mini has no problem. It works very well.

The few models I tried on Groq and here are some observations
- llama3-groq-8b-8192-tool-use-preview (works)
- gemma2-9b-it (works well)
- and llama-3.1-70b-versatile and some others (don't work at all)

I was hoping that llama-3.1-70b-versatile would have worked well. So we could look into running it via Ollama.

My questions
1. I'm seeing in the logs that the tool schema provided to the LLMs is the OpenAI schema which must be because I'm using the OpenAIChatModel.. Should I be using a different one to talk to the other LLMs?

2. Do they all support the OpenAI schema? Which is probably derived or the same as the json-schema.org?

3. Also, looks like the LLMs should inherently support tool calling which I assumed the newer llama-3.1-70b-versatile would do. Is this not the case?

4. Are there any recommendations for equivalent or cheaper (performance/cost) option for tool calling other than gpt-4o-mini?

**Comment by glaforge** [comment]
Maybe the specific version of the Llama model you are using doesn't support function calling indeed?
I was reading this [page](https://groq.com/introducing-llama-3-groq-tool-use-models/) where I see some other model names for Llama 3.1 70B that seem to indicate that function calling can work.
Maybe double check the version name you're using?

**Reply by rwankar** [reply]
Thanks. It looks like Groq has fine tuned the Llama for tool use. I assumed Meta would release models that inherently support tools.

When looking for Llama 3.1 models supporting tools I came across this https://docs.together.ai/docs/llama-3-function-calling which says that 

> Llama 3.1 shipped natively with function calling support, but instead of specifying a tool_choice parameter like traditional function calling, it works with a special prompt syntax.

Does this mean LangChain4j won't work directly with the 3.1 model? That link later says that OpenAI style is also supported. Looks like that's special if I use together.ai. Looks like they've fine tuned it as well?

I'm still trying to get my head around all this.

**Reply by glaforge** [reply]
That's the model that decides how function calling works. So unless you follow its conventions, it won't reply with a function call request. So, as far as I understand, there's not much that LangChain4j can do here, I'm afraid :-(

**Reply by rwankar** [reply]
I'm not sure what LangChain's goals are but given that it offers tools abstraction classes and supports multiple ChatModels I suppose the vision is to provide a uniform interface? Of course it's for them to decide if and how they want to handle it.

I just wanted to if tools were expected to work seamlessly across LLMs and if I was using it right. May be I'll add a feature request and see what the opinion is.
